micro-rdk/src/esp32/dtls.rs

```diff
@@ -139,37 +145,114 @@ extern "C" fn ssl_debug(
     log!(level, "[mbedtls] {}:{} - {}", file, line, msg);
 }
 
+#[derive(Debug)]
+enum DelayState {
+    Empty,
+    Intermediate,
+    Fin,
+}
+
+impl Default for DelayState {
+    fn default() -> Self {
+        Self::Empty
+    }
+}
+
 #[derive(Debug, Default)]
 struct Esp32DtlsDelay {
     intermediate: Option<Instant>,
     fin: Option<Instant>,
+    ssl_ctx: Option<*mut mbedtls_ssl_context>,
+    timer: Option<Timer>,
+    state: DelayState,
+}
+
+impl Esp32DtlsDelay {
+    fn set_delay(&mut self, intermediate: Instant, fin: Instant) {
+        let _ = self.intermediate.insert(intermediate);
+        let _ = self.fin.insert(fin);
+        self.state = DelayState::Intermediate;
+        let _ = std::mem::replace(&mut self.timer, Some(Timer::at(intermediate)));
+    }
+    fn poll_delay(&mut self, cx: &mut Context<'_>) {
+        let timer = self.timer.take();
+        let state = std::mem::replace(&mut self.state, DelayState::Empty);
+        if let Some(mut timer) = timer {
+            match Pin::new(&mut timer).poll(cx) {
+                Poll::Pending => {
+                    self.timer = Some(timer);
+                    self.state = state;
+                }
+                Poll::Ready(_) => match state {
+                    DelayState::Empty => unreachable!(),
+                    DelayState::Intermediate => {
+                        self.timer = Some(Timer::at(*self.fin.as_ref().unwrap()));
+                        self.state = DelayState::Fin
+                    }
+                    DelayState::Fin => {}
+                },
+            }
+        }
+    }
+    fn reset_delay(&mut self) {
+        let _ = self.intermediate.take();
+        let _ = self.fin.take();
+        let _ = self.timer.take();
+    }
+    fn get_delay(&self) -> i32 {
+        if self.fin.is_none() {
+            return -1;
+        }
+        let now = Instant::now();
+        if now > *self.intermediate.as_ref().unwrap() {
+            log::debug!("intermetidate timer expired");
+            return 2;
+        }
+        if now > *self.fin.as_ref().unwrap() {
```

Comment:
shouldn't we check fin before intermediate?

Reply:
nice catch that was really wrong :)